daemon/cluster/cluster.go

```diff
@@ -239,8 +293,16 @@ func (c *Cluster) startNewNode(forceNewCluster bool, listenAddr, joinAddr, secre
 		reconnectDelay: initialReconnectDelay,
 	}
 	c.node = node
+	c.localAddr = localAddr
+	c.remoteAddr = remoteAddr
 	c.listenAddr = listenAddr
+	c.advertiseAddr = advertiseAddr
 	c.saveState()
+
+	// Do this assignment after saveState so that the result of
+	// the resolution won't be saved persistently.
+	c.localAddr = actualLocalAddr
```

Comment:
> On reconnection it seems to be reusing c.localAddr and still saving it.

I've updated the PR to fix this.

I also squashed the commits and implemented @mavenugo's suggestion of filtering all network interfaces that contains an address in any `docker network` subnet.